packages/fuels-core/src/code_gen/flat_abigen.rs

```diff
@@ -0,0 +1,261 @@
+                #![allow(dead_code)]
+                #![allow(unused_imports)]
+
+                #includes
+
+                #code
+
+                #abi_structs
+                #abi_enums
+            }
+        })
+    }
+
+    pub fn functions(&self) -> Result<TokenStream, Error> {
+        let mut tokenized_functions = Vec::new();
+
+        for function in &self.abi.functions {
+            let tokenized_fn = _new_expand_function(function, &self.types)?;
+            tokenized_functions.push(tokenized_fn);
+        }
+
+        Ok(quote! { #( #tokenized_functions )* })
+    }
+
+    fn abi_structs(&self) -> Result<TokenStream, Error> {
+        let mut structs = TokenStream::new();
+
+        // Prevent expanding the same struct more than once
+        let mut seen_struct: Vec<&str> = vec![];
+
+        for prop in &self.abi.types {
+            // If it isn't a struct, skip.
+            if !prop.is_struct_type() {
+                continue;
+            }
+
+            // Skip custom type generation if the custom type is a Sway-native type.
+            // This means ABI methods receiving or returning a Sway-native type
+            // can receive or return that native type directly.
+            if FlatAbigen::is_sway_native_type(&prop.type_field) {
+                continue;
+            }
+
+            if !seen_struct.contains(&prop.type_field.as_str()) {
+                structs.extend(_new_expand_custom_struct(prop, &self.types)?);
+                seen_struct.push(&prop.type_field);
+            }
+        }
+
+        Ok(structs)
+    }
+
+    // Checks whether the given type field is a Sway-native type.
+    // It's expected to come in as `"struct T"` or `"enum T"`.
+    // `T` is a Sway-native type if it matches exactly one of
+    // the reserved strings, such as "Address" or "ContractId".
+    fn is_sway_native_type(type_field: &str) -> bool {
+        let split: Vec<&str> = type_field.split_whitespace().collect();
+
+        if split.len() > 2 {
+            return false;
+        }
+        split[1] == CONTRACT_ID_SWAY_NATIVE_TYPE || split[1] == ADDRESS_SWAY_NATIVE_TYPE
+    }
+
+    fn abi_enums(&self) -> Result<TokenStream, Error> {
+        let mut enums = TokenStream::new();
+
+        for prop in &self.abi.types {
+            if !prop.is_enum_type() {
+                continue;
+            }
+            enums.extend(_new_expand_custom_enum(prop, &self.types)?);
+        }
+
+        Ok(enums)
+    }
+
+    /// Reads the parsed ABI and returns the custom types (either `struct` or `enum`) found in it.
+    /// Custom types can be in the free form (`Struct Person`, `Enum State`), inside arrays (`[struct Person; 2]`, `[enum State; 2]`)), or
+    /// inside tuples (`(struct Person, struct Address)`, `(enum State, enum Country)`).
+    pub fn get_custom_types(abi: &ProgramABI) -> HashMap<usize, TypeDeclaration> {
+        abi.types.iter().map(|t| (t.type_id, t.clone())).collect()
+    }
+}
```

Comment:
The docstring and naming should be updated to reflect that this collects *all* types